tests/framework/prototype.py

```diff
@@ -38,41 +61,82 @@ def search_path(yaml_object, steps, default_value = None):
-    return shell_subprocess_run("kubectl cluster-info --context kind-kind", check = False) == 0
+class KubernetesClusterManager:
+    """
+    KubernetesClusterManager controlls the lifecycle of KinD cluster and Kubernetes API client.
+    """
+    def __init__(self) -> None:
+        self.k8s_client_dict = {}
+
+    def delete_kind_cluster(self) -> None:
+        """Delete a KinD cluster"""
+        shell_subprocess_run("kind delete cluster")
+        for _, k8s_client in self.k8s_client_dict.items():
+            k8s_client.api_client.rest_client.pool_manager.clear()
+            k8s_client.api_client.close()
+        self.k8s_client_dict = {}
+
+    def create_kind_cluster(self, kind_config = None) -> None:
+        """Create a KinD cluster"""
+        # To use NodePort service, `kind_config` needs to set `extraPortMappings` properly.
+        kind_config = CONST.DEFAULT_KIND_CONFIG if not kind_config else kind_config
+        shell_subprocess_run(f"kind create cluster --wait 900s --config {kind_config}")
+        config.load_kube_config()
+        self.k8s_client_dict.update({
+            CONST.K8S_V1_CLIENT_KEY: client.CoreV1Api(),
+            CONST.K8S_CR_CLIENT_KEY: client.CustomObjectsApi()
+        })
+
+    def check_cluster_exist(self) -> bool:
+        """Check whether KinD cluster exists or not"""
+        return shell_subprocess_run("kubectl cluster-info --context kind-kind", check = False) == 0
+K8S_CLUSTER_MANAGER = KubernetesClusterManager()
+
+class OperatorManager:
+    """
+    OperatorManager controlls the lifecycle of KubeRay operator. It will download Docker images,
+    load images into an existing KinD cluster, and install CRD and KubeRay operator.
+    """
+    def __init__(self, docker_image_dict) -> None:
+        for key in [CONST.OPERATOR_IMAGE_KEY, CONST.RAY_IMAGE_KEY]:
+            if key not in docker_image_dict:
+                raise Exception(f"Image {key} does not exist!")
+        self.docker_image_dict = docker_image_dict
+
+    def prepare_operator(self):
+        """Prepare KubeRay operator for an existing KinD cluster"""
+        self.__kind_prepare_images()
+        self.__install_crd_and_operator()
+
+    def __kind_prepare_images(self):
+        """Download images and load images into KinD cluster"""
+        def download_images():
+            """Download Docker images from DockerHub"""
+            logger.info("Download Docker images: %s", self.docker_image_dict)
+            for key in self.docker_image_dict:
+                # Only pull the image from DockerHub when the image does not
+                # exist in the local docker registry.
+                image = self.docker_image_dict[key]
+                if shell_subprocess_run(
+                        f'docker image inspect {image} > /dev/null', check = False) != 0:
+                    shell_subprocess_run(f'docker pull {image}')
+                else:
+                    logger.info("Image %s exists", image)
+
+        download_images()
```

Comment:
I am wondering why you define the inner function download_images() here instead of writing them down directly.

Reply:
We always load Docker images into a KinD cluster after `download_images()`. Hence, defining `download_images()` as an inner function can reduce the cognitive load on developers who use the module. To elaborate, they did not need to know that `kind load ...` must be called after `download_images()`.

Reply:
It's a good chance to read Chapter 5 of the book "A Philosophy of Software Design".